File: tests/instructions/createAuctionHouse.ts
import { Program, Wallet, Provider } from "@coral-xyz/anchor";

import { AuctionHouse } from "../../target/types/auction_house";
import { Auctioneer } from "../../target/types/auctioneer";
import { NATIVE_MINT } from "@solana/spl-token";

import * as pda from "../pda";
import { createSystemAccount } from "../utils";
import { AuctionHouseData } from "../interfaces";

export async function createAuctionHouse({
  auctionHouseProgram,
  auctioneerProgram,
  provider,
  payer,
}: {
  auctionHouseProgram: Program<AuctionHouse>;
  auctioneerProgram: Program<Auctioneer>;
  provider: Provider;
  payer: Wallet;
}) {
  const authorityKeypair = await createSystemAccount({ provider, payer });

  const treasuryMint = NATIVE_MINT;

  const [auctionHouseAddress, auctionHouseBump] = pda.findAuctionHouseAddress({
    authority: authorityKeypair.publicKey,
    mint: treasuryMint,
  });
  const [auctionHouseFeeAccountAddress, feePayerBump] =
    pda.findAuctionHouseFeeAccountAddress({ auctionHouseAddress });
  const [auctionHouseTreasuryAddress, treasuryBump] =
    pda.findAuctionHouseTreasuryAddress({ auctionHouseAddress });
  const sellerFeeBasisPoints = 100;
  const canChangeSalePrice = false;

  const createAuctionHouseTx = await auctionHouseProgram.methods
    .createAuctionHouse(
      auctionHouseBump,
      feePayerBump,
      treasuryBump,
      sellerFeeBasisPoints,
      canChangeSalePrice
    )
    .accounts({
      treasuryMint,
      payer: authorityKeypair.publicKey, // Signer
      authority: authorityKeypair.publicKey,
      feeWithdrawalDestination: payer.publicKey,
      treasuryWithdrawalDestination: payer.publicKey,
      treasuryWithdrawalDestinationOwner: payer.publicKey,
      auctionHouse: auctionHouseAddress,
      auctionHouseFeeAccount: auctionHouseFeeAccountAddress,
      auctionHouseTreasury: auctionHouseTreasuryAddress,
    })
    .signers([authorityKeypair])
    .rpc();
  console.log("Transaction [Create Auction House]", createAuctionHouseTx);

  const [auctioneerAuthorityAddress] = pda.findAuctioneerAuthorityAddress({
    auctionHouseAddress,
  });
  const [auctioneerAddress] = pda.findAuctioneerAddress({
    auctionHouseAddress,
    auctioneerAuthorityAddress,
  });

  const delegateAuctioneerTx = await auctionHouseProgram.methods
    .delegateAuctioneer()
    .accounts({
      auctionHouse: auctionHouseAddress,
      authority: authorityKeypair.publicKey, // Signer
      auctioneerAuthority: auctioneerAuthorityAddress,
      auctioneer: auctioneerAddress,
    })
    .signers([authorityKeypair])
    .rpc();
  console.log("Transaction [Delegate Auctioneer]", delegateAuctioneerTx);

  const authorizeAuctioneerTx = await auctioneerProgram.methods
    .authorize()
    .accounts({
      wallet: authorityKeypair.publicKey, // Signer
      auctionHouse: auctionHouseAddress,
      auctioneerAuthority: auctioneerAuthorityAddress,
    })
    .signers([authorityKeypair])
    .rpc();
  console.log("Transaction [Authorize Auctioneer]", authorizeAuctioneerTx);

  const ah = await auctionHouseProgram.account.auctionHouse.fetch(
    auctionHouseAddress
  );

  const auctionHouse: AuctionHouseData = {
    address: auctionHouseAddress,
    authorityKeypair: authorityKeypair,
    auctionHouseFeeAccount: ah.auctionHouseFeeAccount,
    auctionHouseTreasury: ah.auctionHouseTreasury,
    treasuryWithdrawalDestination: ah.treasuryWithdrawalDestination,
    feeWithdrawalDestination: ah.feeWithdrawalDestination,
    treasuryMint: ah.treasuryMint,
    authority: ah.authority,
    creator: ah.creator,
    bump: ah.bump,
    treasuryBump: ah.treasuryBump,
    feePayerBump: ah.feePayerBump,
    sellerFeeBasisPoints: ah.sellerFeeBasisPoints,
    canChangeSalePrice: ah.canChangeSalePrice,
    escrowPaymentBump: ah.escrowPaymentBump,
    hasAuctioneer: ah.hasAuctioneer,
    auctioneerAddress: ah.auctioneerAddress,
  };

  return auctionHouse;
}

File: programs/auction/tests/create_auction_house.rs
use anchor_client::solana_sdk::{signature::Keypair, signer::Signer, transaction::Transaction};
use anchor_lang::{
    solana_program::{instruction::Instruction, system_program, sysvar},
    AccountDeserialize, InstructionData, ToAccountMetas,
};
use anchor_spl::associated_token;
use auction_house::{pda::*, AuctionHouse};
use solana_program_test::tokio;

mod utils;
use utils::setup_functions::*;

#[tokio::test]
async fn create_auction_house_success() {
    let mut context = auction_house_program_test().start_with_context().await;

    let authority = Keypair::new();
    airdrop(&mut context, &authority.pubkey(), 10 * ONE_SOL)
        .await
        .unwrap();

    let treasury_mint = spl_token::native_mint::id();

    let (auction_house, auction_house_bump) =
        find_auction_house_address(&authority.pubkey(), &treasury_mint);
    let (auction_house_fee_account, auction_house_fee_account_bump) =
        find_auction_house_fee_account_address(&auction_house);
    let (auction_house_treasury, auction_house_treasury_bump) =
        find_auction_house_treasury_address(&auction_house);

    let data = auction_house::instruction::CreateAuctionHouse {
        _bump: auction_house_bump,
        fee_payer_bump: auction_house_fee_account_bump,
        treasury_bump: auction_house_treasury_bump,
        seller_fee_basis_points: 10,
        can_change_sale_price: false,
    };

    let accounts = auction_house::accounts::CreateAuctionHouse {
        treasury_mint,
        payer: authority.pubkey(),
        authority: authority.pubkey(),
        fee_withdrawal_destination: context.payer.pubkey(),
        treasury_withdrawal_destination: context.payer.pubkey(),
        treasury_withdrawal_destination_owner: context.payer.pubkey(),
        auction_house,
        auction_house_fee_account,
        auction_house_treasury,
        token_program: spl_token::id(),
        system_program: system_program::id(),
        associated_token_program: associated_token::ID,
        rent: sysvar::rent::id(),
    };

    let ix = Instruction {
        program_id: auction_house::id(),
        data: data.data(),
        accounts: accounts.to_account_metas(None),
    };

    let tx = Transaction::new_signed_with_payer(
        &[ix],
        Some(&authority.pubkey()),
        &[&authority],
        context.last_blockhash,
    );
    context.banks_client.process_transaction(tx).await.unwrap();

    let auction_house_account = context
        .banks_client
        .get_account(auction_house)
        .await
        .unwrap()
        .expect("Auction House account not found");

    let auction_house_data =
        AuctionHouse::try_deserialize(&mut auction_house_account.data.as_ref()).unwrap();

    assert_eq!(auction_house_data.authority, authority.pubkey());
}

#[tokio::test]
async fn failure_create_auction_house_reinitialization() {
    let mut context = auction_house_program_test().start_with_context().await;

    // CreateAuctionHouse

    let authority = Keypair::new();
    airdrop(&mut context, &authority.pubkey(), 10 * ONE_SOL)
        .await
        .unwrap();

    let treasury_mint = spl_token::native_mint::id();

    let (auction_house, auction_house_bump) =
        find_auction_house_address(&authority.pubkey(), &treasury_mint);
    let (auction_house_fee_account, auction_house_fee_account_bump) =
        find_auction_house_fee_account_address(&auction_house);
    let (auction_house_treasury, auction_house_treasury_bump) =
        find_auction_house_treasury_address(&auction_house);

    let data = auction_house::instruction::CreateAuctionHouse {
        _bump: auction_house_bump,
        fee_payer_bump: auction_house_fee_account_bump,
        treasury_bump: auction_house_treasury_bump,
        seller_fee_basis_points: 10,
        can_change_sale_price: false,
    };

    let accounts = auction_house::accounts::CreateAuctionHouse {
        treasury_mint,
        payer: authority.pubkey(),
        authority: authority.pubkey(),
        fee_withdrawal_destination: context.payer.pubkey(),
        treasury_withdrawal_destination: context.payer.pubkey(),
        treasury_withdrawal_destination_owner: context.payer.pubkey(),
        auction_house,
        auction_house_fee_account,
        auction_house_treasury,
        token_program: spl_token::id(),
        system_program: system_program::id(),
        associated_token_program: associated_token::ID,
        rent: sysvar::rent::id(),
    };

    let ix = Instruction {
        program_id: auction_house::id(),
        data: data.data(),
        accounts: accounts.to_account_metas(None),
    };

    let tx = Transaction::new_signed_with_payer(
        &[ix],
        Some(&authority.pubkey()),
        &[&authority],
        context.last_blockhash,
    );
    context.banks_client.process_transaction(tx).await.unwrap();

    // Attempt to reinitialize (Hack)

    let malicious_wallet = Keypair::new();
    airdrop(&mut context, &malicious_wallet.pubkey(), 10 * ONE_SOL)
        .await
        .unwrap();

    let hacked_fee_withdrawal_destination = malicious_wallet.pubkey();
    let hacked_treasury_withdrawal_destination = malicious_wallet.pubkey();
    let hacked_treasury_withdrawal_destination_owner = malicious_wallet.pubkey();

    let data = auction_house::instruction::CreateAuctionHouse {
        _bump: auction_house_bump,
        fee_payer_bump: auction_house_fee_account_bump,
        treasury_bump: auction_house_treasury_bump,
        seller_fee_basis_points: 10,
        can_change_sale_price: false,
    };

    let accounts = auction_house::accounts::CreateAuctionHouse {
        treasury_mint,
        payer: authority.pubkey(),
        authority: authority.pubkey(),
        fee_withdrawal_destination: hacked_fee_withdrawal_destination,
        treasury_withdrawal_destination: hacked_treasury_withdrawal_destination,
        treasury_withdrawal_destination_owner: hacked_treasury_withdrawal_destination_owner,
        auction_house,
        auction_house_fee_account,
        auction_house_treasury,
        token_program: spl_token::id(),
        system_program: system_program::id(),
        associated_token_program: associated_token::ID,
        rent: sysvar::rent::id(),
    };

    let ix = Instruction {
        program_id: auction_house::id(),
        data: data.data(),
        accounts: accounts.to_account_metas(None),
    };

    let tx = Transaction::new_signed_with_payer(
        &[ix],
        Some(&authority.pubkey()),
        &[&authority],
        context.last_blockhash,
    );

    let tx_error = context
        .banks_client
        .process_transaction(tx)
        .await
        .unwrap_err();

    assert_error(tx_error, ERR_AUCTION_HOUSE_ALREADY_INITIALIZED);
}
